File: onto-code-maker-result-tester/src/main/java/ontology/generator/classes/examples/family/serialization/WomanSerialization.java

package ontology.generator.classes.examples.family.serialization;

import org.eclipse.rdf4j.model.*;
import java.util.Collection;
import java.util.Set;
import java.util.HashSet;
import java.util.*;

import java.util.stream.Collectors;
import ontology.generator.classes.examples.family.entities.*;
import org.eclipse.rdf4j.model.util.Values;
import org.eclipse.rdf4j.model.vocabulary.RDF;

import java.time.LocalTime;
import ontology.generator.classes.examples.family.Vocabulary;
import ontology.generator.classes.examples.family.entities.WomanInt;
import ontology.generator.classes.examples.family.entities.Men;
import ontology.generator.classes.examples.family.entities.Men;

public class WomanSerialization extends SerializationModel<WomanInt>{

    @Override
    public void addToModel(Model model, WomanInt woman) {
        model.add(woman.getIri(),RDF.TYPE, woman.getClassIRI());
        addPropertiesToModel(model,woman);

    }

    protected void addPropertiesToModel(Model model,  WomanInt woman) {
        List<OntoEntity> hasHusbandPom = new ArrayList<>();
        hasHusbandPom.addAll(woman.getHasHusband());
        setRDFCollection(model,woman.getIri(),Vocabulary.HASHUSBAND_PROPERTY_IRI,hasHusbandPom);

        if(woman.getMen() != null ){
            model.add(woman.getIri(),Vocabulary.MEN_PROPERTY_IRI,woman.getMen().getIri());
        }


         new HumanSerialization().addPropertiesToModel(model, woman);
    }

    protected void setProperties(Model model,WomanInt woman,int nestingLevel) throws Exception{
        Set<Resource> hasHusband = super.getAllResourceObjects(model,Vocabulary.HASHUSBAND_PROPERTY_IRI,woman.getIri());
        // add also all values from inverse property hasWife
         hasHusband.addAll(super.getAllIRISubjects(model,Vocabulary.HASWIFE_PROPERTY_IRI,woman.getIri()));
        for(Resource propValue:hasHusband){
            if(propValue.isIRI()) {
                Men hasHusbandInstance = new MenSerialization().getInstanceFromModel(model, (IRI) propValue,nestingLevel);
                if(hasHusbandInstance == null) throw new Exception("Instance of " + propValue.stringValue() + " is not in model.");
                woman.addHasHusband(hasHusbandInstance);
            }else if(propValue.isBNode()){
                List<Value> listOfValues = super.getRDFCollection(model,(BNode)propValue);
                for(Value value:listOfValues){
                    if(value.isIRI()){
                        Men hasHusbandInstance = new MenSerialization().getInstanceFromModel(model, (IRI)value,nestingLevel);
                        if(hasHusbandInstance == null) throw new Exception("Instance of " + propValue.stringValue() + " is not in model.");
                        woman.addHasHusband(hasHusbandInstance);
                    }
                 }
            }
        }
        IRI men = super.getFirstIriObject(model,Vocabulary.MEN_PROPERTY_IRI,woman.getIri());
        if ( men == null){
            // check inverse functional property
            men = super.getSubjectOfCollectionValue(model,Vocabulary.HASWIFE_PROPERTY_IRI,woman.getIri());
        }
        if ( men != null ){
            Men menInstance = new MenSerialization().getInstanceFromModel(model, men,nestingLevel);
            woman.setMen(menInstance);
        }

        new HumanSerialization().setProperties(model, woman,nestingLevel);
    }

    @Override
    public WomanInt getInstanceFromModel(Model model,IRI instanceIri,int nestingLevel) throws Exception{
        Model statements = model.filter(instanceIri,RDF.TYPE,Woman.CLASS_IRI);
        if(statements.size() != 0){
            Woman woman = new Woman(instanceIri);
            if(nestingLevel > 0){
                nestingLevel--;
                setProperties(model, woman,nestingLevel);
            }
            return woman;
        }

        return null;
    }

    @Override
    public Collection<WomanInt> getAllInstancesFromModel(Model model,int nestingLevel)throws Exception{
        Model statements = model.filter(null,RDF.TYPE,Woman.CLASS_IRI);
        Collection<WomanInt> allInstances = new ArrayList<>();
        for(Statement statement:statements){
            Resource subject = statement.getSubject();
            if(subject.isIRI()){
                IRI iri = (IRI) subject;
                WomanInt woman = getInstanceFromModel(model,iri,nestingLevel);
                allInstances.add(woman);
            }
        }

        return allInstances;
    }

    @Override
    public void removeInstanceFromModel(Model model,IRI instanceIri) {
        Set<Value> rdfCollections = model.filter(instanceIri,null,null).objects();
        List<Value> listOfnodes = rdfCollections.stream().filter(o -> o.isBNode()).collect(Collectors.toList());
        for(Value node:listOfnodes){
            model.removeAll(getModelRDFCollection(model,(BNode)node));
            model.remove(instanceIri,null,(BNode)node);
        }
        model.remove(instanceIri,RDF.TYPE,Woman.CLASS_IRI);
        model.remove(instanceIri,null,null);
    }

    @Override
    public void updateInstanceInModel(Model model,WomanInt woman){
        Set<Value> rdfCollections = model.filter(woman.getIri(),null,null).objects();
        List<Value> listOfnodes = rdfCollections.stream().filter(Value::isBNode).collect(Collectors.toList());
        for(Value node:listOfnodes){
            model.removeAll(getModelRDFCollection(model,(BNode)node));
            model.remove(woman.getIri(),null,node);
        }

        Model statements = model.filter(woman.getIri(),null,null);
        statements.removeIf(x -> !x.getPredicate().equals(RDF.TYPE));

        addPropertiesToModel(model,woman);
    }



}

File: onto-code-maker-result-tester/src/test/python/familyPython/serialization/WomanSerialization.py

from rdflib.graph import URIRef, Literal,BNode,Collection
from rdflib.namespace import RDF
from Vocabulary import Vocabulary
from entities.Woman import Woman
from .SerializationModel import SerializationModel

class WomanSerialization(SerializationModel):

    def addToModel(self,ontology, woman):
        ontology.add((woman.getIri(),RDF.type, woman.getClassIRI()))
        self.addPropertiesToModel(ontology,woman)


    def addPropertiesToModel(self,ontology, woman):
        hasHusbandPom = set(woman.getHasHusband())
        super().setRDFCollection(ontology,woman.getIri(),Vocabulary.HASHUSBAND_PROPERTY_IRI,hasHusbandPom)
        if woman.getMen() is not None:
            ontology.add((woman.getIri(),Vocabulary.MEN_PROPERTY_IRI,woman.getMen().getIri()))

        from .HumanSerialization import HumanSerialization
        HumanSerialization().addPropertiesToModel(ontology, woman)
        pass

    def setProperties(self,ontology, woman, nestingLevel):
        hasHusband = super().getAllResourceObjects(ontology,Vocabulary.HASHUSBAND_PROPERTY_IRI,woman.getIri())
        # add also all values from inverse property hasWife
        for p in super().getAllIRISubjects(ontology,Vocabulary.HASWIFE_PROPERTY_IRI,woman.getIri()):
            hasHusband.add(p)
        for propValue in hasHusband:
            if isinstance(propValue,URIRef):
                from .MenSerialization import MenSerialization
                hasHusbandInstance = MenSerialization().getInstanceFromModel(ontology, propValue,nestingLevel)
                if hasHusbandInstance is None:
                    print("Instance of " + str(propValue) + " is not in model.")
                woman.addHasHusband(hasHusbandInstance)
            elif isinstance(propValue,BNode):
                listOfValues = super().getRDFCollection(ontology,propValue)
                for value in listOfValues:
                    if isinstance(value,URIRef):
                        from .MenSerialization import MenSerialization
                        hasHusbandInstance = MenSerialization().getInstanceFromModel(ontology,value,nestingLevel)
                        if hasHusbandInstance is None:
                            print("Instance of " + str(propValue) + " is not in model.")
                        woman.addHasHusband(hasHusbandInstance)
        men = super().getFirstIriObject(ontology,Vocabulary.MEN_PROPERTY_IRI,woman.getIri())
        if men is None:
            # check inverse functional property
            men = super().getSubjectOfCollectionValue(ontology,Vocabulary.HASWIFE_PROPERTY_IRI,woman.getIri())
        if men is not None:
            from .MenSerialization import MenSerialization
            menInstance = MenSerialization().getInstanceFromModel(ontology, men,nestingLevel)
            woman.setMen(menInstance)

        from .HumanSerialization import HumanSerialization
        HumanSerialization().setProperties(ontology, woman,nestingLevel)
        pass

    def getInstanceFromModel(self,ontology, instanceIri,nestingLevel):
        if (instanceIri,RDF.type,Woman.CLASS_IRI) in ontology:
            woman =  Woman(instanceIri)
            if nestingLevel > 0:
                nestingLevel -= 1
                self.setProperties(ontology, woman,nestingLevel)
            return woman
        return None

    def getAllInstancesFromModel(self,ontology,nestingLevel):
        allInstances = set()
        for s, p, o in ontology.triples((None,RDF.type,Woman.CLASS_IRI)):
            if isinstance(s,URIRef) :
                woman = self.getInstanceFromModel(ontology,s,nestingLevel)
                allInstances.add(woman)
        return allInstances

    def removeInstanceFromModel(self,ontology, subject):
        ontology.remove((subject,RDF.type,Woman.CLASS_IRI))
        for statement in ontology.triples((subject,None,None)):
            ontology.remove(statement)



